src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1914 @@
+        } else if (si->si_status != EXIT_SUCCESS)
+                if (j->status_errno != 0)
+                        sd_bus_error_set_errno(&error, j->status_errno);
+                else
+                        sd_bus_error_setf(&error, SD_BUS_ERROR_FAILED,
+                                          "Job failed with exit code %i.", si->si_status);
+        else {
+                r = job_parse_child_output(TAKE_FD(j->stdout_fd), &json);
+                if (r < 0)
+                        sd_bus_error_set_errnof(&error, r, "Failed to parse JSON: %m");
+        }
+
+        /* Only send notification of exit if the job was actually detached */
+        if (j->detach_cb) {
+                r = sd_bus_emit_signal(
+                                j->manager->bus,
+                                "/org/freedesktop/sysupdate1",
+                                "org.freedesktop.sysupdate1.Manager",
+                                "JobRemoved",
+                                "toi",
+                                j->id,
+                                j->object_path,
+                                j->status_errno != 0 ? -j->status_errno : si->si_status);
+                if (r < 0)
+                        log_warning_errno(r, "Cannot emit JobRemoved message, ignoring: %m");
+        }
+
+        if (j->dbus_msg && j->complete_cb) {
+                if (sd_bus_error_is_set(&error)) {
+                        log_warning("Bus error occurred, ignoring callback for job: %s", error.message);
+                        sd_bus_reply_method_error(j->dbus_msg, &error);
+                } else {
+                        r = j->complete_cb(j->dbus_msg, j, json, &error);
+                        if (r < 0) {
+                                log_warning_errno(r, "Error during execution of job callback: %s", bus_error_message(&error, r));
+                                sd_bus_reply_method_errno(j->dbus_msg, r, &error);
+                        }
+                }
+        }
+
+        if (j->manager && hashmap_isempty(j->manager->jobs))
+                manager_became_idle(j->manager);
+
+        job_free(j);
+
+        return 0;
+}
+
+static inline const char* sysupdate_binary_path(void) {
+        return secure_getenv("SYSTEMD_SYSUPDATE_PATH") ?: SYSTEMD_SYSUPDATE_PATH;
+}
+
+static int target_get_argument(Target *t, char **ret) {
+        _cleanup_free_ char *target_arg = NULL;
+
+        assert(t);
+        assert(ret);
+
+        if (t->class != TARGET_HOST) {
+                if (t->class == TARGET_COMPONENT)
+                        target_arg = strjoin("--component=", t->name);
+                else if (IN_SET(t->image_type, IMAGE_DIRECTORY, IMAGE_SUBVOLUME))
+                        target_arg = strjoin("--root=", t->path);
+                else if (IN_SET(t->image_type, IMAGE_RAW, IMAGE_BLOCK))
+                        target_arg = strjoin("--image=", t->path);
+                else
+                        assert_not_reached();
```

Comment:
doesn't matter, but probably a job for a switch statement

Reply:
I'm not sure about this one: we check `t->class` against `TARGET_HOST` and `TARGET_COMPONENT` then check `t->image_type` against a few `IMAGE_*`.

How should we do it? two switches? a switch in the else?